crates/engine_bibtex/src/c_api/exec.rs

```diff
@@ -1,367 +1,285 @@
             write_logs("`");
-            match print_a_pool_str(val.lit as usize) {
-                CResult::Ok => (),
-                err => return err,
-            }
+            rs_print_a_pool_str(s, pool)?;
             write_logs("` is a missing field");
-            CResult::Ok
         }
-        StkType::Illegal => {
+        ExecVal::Illegal => {
             illegl_literal_confusion();
-            CResult::Error
+            return Err(BibtexError::Fatal);
         }
     }
+    Ok(())
 }
 
-#[no_mangle]
-pub unsafe extern "C" fn print_wrong_stk_lit(
-    ctx: *mut ExecCtx,
+pub(crate) fn print_wrong_stk_lit(
+    ctx: &mut ExecCtx,
+    pool: &StringPool,
+    hash: &HashData,
     val: ExecVal,
     typ2: StkType,
-) -> CResult {
-    if val.typ != StkType::Illegal {
-        match print_stk_lit(val) {
-            CResult::Ok => (),
-            err => return err,
-        }
-        let res = match typ2 {
-            StkType::Integer => {
-                write_logs(", not an integer,");
-                CResult::Ok
-            }
-            StkType::String => {
-                write_logs(", not a string,");
-                CResult::Ok
-            }
-            StkType::Function => {
-                write_logs(", not a function,");
-                CResult::Ok
-            }
-            StkType::Missing | StkType::Illegal => {
-                illegl_literal_confusion();
-                CResult::Error
-            }
-        };
-        match res {
-            CResult::Ok => (),
-            err => return err,
+) -> Result<(), BibtexError> {
+    match val {
+        ExecVal::Illegal => Ok(()),
+        _ => {
+            print_stk_lit(val, pool, hash)?;
+
+            match typ2 {
+                StkType::Integer => write_logs(", not an integer,"),
+                StkType::String => write_logs(", not a string,"),
+                StkType::Function => write_logs(", not a function,"),
+                StkType::Missing | StkType::Illegal => {
+                    illegl_literal_confusion();
+                    return Err(BibtexError::Fatal);
+                }
+            };
+
+            bst_ex_warn_print(ctx, pool)
         }
-        rs_bst_ex_warn_print(&*ctx).into()
-    } else {
-        CResult::Ok
     }
 }
 
-pub fn rs_bst_ex_warn_print(ctx: &ExecCtx) -> Result<(), BibtexError> {
+pub fn bst_ex_warn_print(ctx: &ExecCtx, pool: &StringPool) -> Result<(), BibtexError> {
     if ctx.mess_with_entries {
         write_logs(" for entry ");
-        let res = with_cites(|ci| print_a_pool_str(ci.get_cite(ci.ptr())));
-        if res != CResult::Ok {
-            return res.into();
-        }
+        with_cites(|ci| rs_print_a_pool_str(ci.get_cite(ci.ptr()), pool))?;
     }
 
     write_logs("\nwhile executing-");
     // SAFETY: The `Bibtex` value pointed to by `ExecCtx` is guaranteed valid
-    bst_ln_num_print(unsafe { &*ctx.glbl_ctx })?;
+    bst_ln_num_print(unsafe { &*ctx.glbl_ctx }, pool)?;
```

Comment:
Could constructs like this be replaced with `ctx.glbl_ctx()` (or `.glbl_ctx_mut()`) now?

Reply:
It should be able to, at this point, I'll do a cleanup to replace the pointer derefs with them.